README.md
PitchBlack
===========

Intro
---

Survival is now far more difficult as the Biters play the relentless alien wildlife truly hellbent on overwhelming you.
Spend your daylight wisely as the biters remain largely dormant unless provoked. But even a daytime provocation is nothing compared to the swarm that awakens at sunset.
Summary

Presently, your first day is bright and will last 2 hours. After that nights will occur lasting roughly 20-40 minutes to 30-40 minutes day time, although days will still be dark.

Biter behavior comes in phases, where they will be more passive during the daytime and become more aggressive at night. Also, their evolution rate will spike during nighttime and diminish during day time.

Keep an eye out for warnings on how to deal with their immediate threat.

Mod settings
---

There are ingame settings to change the difficulty (Options -> Mods -> Startup / Map)
For descriptions hover over the names

Complimenting Mods
---

Increased Rates - https://mods.factorio.com/mods/reapa/IncreasedRates
Armageddon - https://mods.factorio.com/mods/binbinhfr/Armageddon
Rampant AI - https://mods.factorio.com/mods/Veden/Rampant


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.0.6
Date: 24. 01. 2019
  Changes:
    - Fixed loading errer
    - Added Russian locale (by badway)
    - Edited locale files
---------------------------------------------------------------------------------------------------
Version: 1.0.5
Date: 23. 01. 2019
  Changes:
    - Add new modifier
---------------------------------------------------------------------------------------------------
Version: 1.0.4
Date: 22. 05. 2018
  Changes:
    - Original data updates and translations
---------------------------------------------------------------------------------------------------
Version: 1.0.3
Date: 15. 03. 2018
  Changes:
    - Balance patched
---------------------------------------------------------------------------------------------------
Version: 1.0.2
Date: 13. 01. 2018
  Bugfixes:
    - Fixed loading errer 
---------------------------------------------------------------------------------------------------
Version: 1.0.1
Date: 06. 01. 2018
  Changes:
    - Updated for Factorio 0.16
---------------------------------------------------------------------------------------------------
Version: 0.2.7
Date: 07. 11. 2017
  Changes:
    - First released

control.lua
require 'Libs/PitchBlack/main'


--On load game for first time
script.on_init(function()
    Main:On_Init()
end)

--On mod version different or if mod did not previously exist
script.on_configuration_changed(function(data)
    if data.mod_changes and data.mod_changes.PitchBlack then
        if not data.mod_changes.PitchBlack.old_version or data.mod_changes.PitchBlack.old_version < '0.2.5' then
            if settings.global["pitch-BiterDamageModifier"] then
                game.forces.enemy.set_ammo_damage_modifier('melee', settings.global["pitch-BiterDamageModifier"].value)
            end
            if settings.global["pitch-SpitterDamageModifier"] then
                game.forces.enemy.set_ammo_damage_modifier('biological', settings.global["pitch-SpitterDamageModifier"].value)
            end
        end
    end
    Main:On_Configuration_Changed(data)
end)

--Called once per in-game tick
--local smoke = {"fire-smoke", "fire-smoke-without-glow", "soft-fire-smoke"}
--local max = 0
--local triggered = false
script.on_event(defines.events.on_tick, function(event)
    --    local count
    --    for _, name in pairs(smoke) do
    --        count = game.surfaces.nauvis.count_entities_filtered{type="smoke"}
    --        if count > max then max = count end
    --        if count and count > 0 then
    --            if not triggered then triggered = true end
    --            log(string.format("%s: %s", name, count))
    --        end
    --        if count == 0 and triggered then
    --            log("max: " .. max)
    --            triggered = false
    --        end
    --    end
    Main:On_Tick(event)
end)

script.on_event(defines.events.on_biter_base_built, 
	function(event)
		Main:On_BiterBuild(event)
	end
)

script.on_event(defines.events.on_runtime_mod_setting_changed, function(event)
    local _, err = pcall(function()
        if event.setting == "pitch-DebugMode" then
            pbLogger.debug_mode = settings.global["pitch-DebugMode"].value
        end
        if event.setting == "pitch-DayLength" then
            Time.DayLength = settings.global["pitch-DayLength"].value
        end
        if event.setting == "pitch-FirstDayPhaseLength" then
            Time.DayPhaseConfig.VariableData[1].MinLength = settings.global["pitch-FirstDayPhaseLength"].value
            Time.DayPhaseConfig.VariableData[1].MaxLength = settings.global["pitch-FirstDayPhaseLength"].value
        end

        if event.setting == "pitch-BiterDamageModifier" then
            game.forces.enemy.set_ammo_damage_modifier('melee', settings.global["pitch-BiterDamageModifier"].value)
        end
        if event.setting == "pitch-SpitterDamageModifier" then
            game.forces.enemy.set_ammo_damage_modifier('biological', settings.global["pitch-SpitterDamageModifier"].value)
        end
    end)
    if err then
        game.print("Pitch Black: Error occured, see log")
        log(serpent.block(err))
    end
end)

script.on_event(defines.events.on_player_died, function(event)
    local _, _ = pcall(function()
        local player = game.players[event.player_index]
        player.surface.create_entity({name = "pitch_explosion", position=player.position})
    end)
end)

remote.add_interface("PitchBlack", 
{
	SkipTimePhase = 
	function() 
		Time.SkipPhase()
	end
})

settings.lua
data:extend{
	{
		--Desc: Max evolution value
		--Default: 100 (100%)
		type = "int-setting",
		name = "pitch-MaxEvolution",
		setting_type = "runtime-global",
		default_value = 100,
		minimum_value = 0,
		order = "pitch-a[Evolution]-a[MaxEvolution]"
	},
	{
		--Desc: Biter expansion scale
		--Default: 100 (100%)
		type = "int-setting",
		name = "pitch-ExpansionScale",
		setting_type = "runtime-global",
		default_value = 100,
		minimum_value = 0,
		maximum_value = 100,
		order = "pitch-a[Evolution]-b[ExpansionScale]"
	},
	{
		--Desc: Biter expansion scale
		--Default: 100 (100%)
		type = "double-setting",
		name = "pitch-ScalePollutionDamage",
		setting_type = "runtime-global",
		default_value = 1.0,
		minimum_value = 0.0,
		maximum_value = 1.0,
		order = "pitch-a[Evolution]-c[ScalePollutionDamage]"
	},
	{
		--Desc: Scalar value for how much the enemy evolves, must be greater than 0.
		--Default: 1 (100%)
		--Note: To make the enemies easier, use a smaller value. To make them harder, use a higher value.
		--ValType: Floating point
		type = "double-setting",
		name = "pitch-ScaleEvolutionRate",
		setting_type = "runtime-global",
		default_value = 1,
		minimum_value = 0.0001,
		order = "pitch-a[Evolution]-d[ScaleEvolutionRate]"
	},
	{
		type = "double-setting",
		name = "pitch-BiterDamageModifier",
		setting_type = "runtime-global",
		default_value = 0,
		minimum_value = -0.9,
		order = "pitch-a[Evolution]-e[BiterDamageModifier]"
	},
	{
		type = "double-setting",
		name = "pitch-SpitterDamageModifier",
		setting_type = "runtime-global",
		default_value = 0,
		minimum_value = -0.9,
		order = "pitch-a[Evolution]-f[SpitterDamageModifier]"
	},
	{
		--Desc: Number of real seconds in a game day
		--Default: 600 (10 minutes)
		--Note: A single day is not necessarily the cycle of the sun... E.G there could be multiple days with the sun up.
		type = "int-setting",
		name = "pitch-PhaseLength",
		setting_type = "runtime-global",
		default_value = 600,
		minimum_value = 1,
		order = "pitch-b[DayLength]-a[DayLength]"
	},
	{
		--Desc: Debug mode, affects things like how verbose logging is to file.
		--Default: false
		type = "bool-setting",
		name = "pitch-DebugMode",
		setting_type = "runtime-global",
		default_value = false,
		order = "pitch-z[Debug]-a[Debug]"
	},
	{

		--Desc: Scalar value for enemy health, must be greater than 0.
		--Default: 1 (100%)
		--Note: To make the enemies weaker, use a smaller value. To make them tougher, use a higher value.
		type = "double-setting",
		name = "pitch-EnemyHealthScale",
		setting_type = "startup",
		default_value = 2.75,
		minimum_value = 0.0001,
		order = "pitch-a[Evolution]-a[EnemyHealthScale]"
	},
	{
		--Desc: Scalar value for how much the enemy spawns, must be greater than 0.
		--Default: 1 (100%)
		--Note: To make the enemies easier, use a smaller value. To make them harder, use a higher value.
		--ValType: Floating point
		--EnemySwarmScale = 1.0,
		type = "double-setting",
		name = "pitch-EnemySwarmScale",
		setting_type = "startup",
		default_value = 1,
		minimum_value = 0.0001,
		order = "pitch-a[Evolution]-b[EnemySwarmScale]"
	},
	{
		type = "double-setting",
		name = "pitch-BiterDamageScale",
		setting_type = "startup",
		default_value = 2.5,
		minimum_value = 0.0001,
		order = "pitch-a[Evolution]-c[BiterDamage]"
	},
	{
		type = "double-setting",
		name = "pitch-SpitterDamageScale",
		setting_type = "startup",
		default_value = 2,
		minimum_value = 0.0001,
		order = "pitch-a[Evolution]-d[SpitterDamage]"
	},
	{
		type = "double-setting",
		name = "pitch-EnemyMovementScale",
		setting_type = "startup",
		default_value = 1,
		minimum_value = 0.0001,
		order = "pitch-a[Evolution]-e[EnemyMovementScale]"
	},
	{
		type = "bool-setting",
		name = "pitch-addResistances",
		setting_type = "startup",
		default_value = true,
		order = "pitch-a[Evolution]-f[addResistances]"
	},
	{
		type = "bool-setting",
		name = "pitch-disableSmoke",
		setting_type = "startup",
		default_value = true,
		order = "pitch-z[disableSmoke]-a[disableSmoke]"
	}
}

local biters = {'small-biter', 'medium-biter', 'big-biter', 'behemoth-biter'}
local spitters = {'small-spitter', 'medium-spitter', 'big-spitter', 'behemoth-spitter'}

for i, name in pairs(biters) do
	data:extend({
		{
			type = "double-setting",
			name = "pitch-" .. name .. "-resistance",
			setting_type = "startup",
			default_value = 10*(i-1)+5,
			minimum_value = 0,
			order = "pitch-a[Evolution]-f[addResistances]-a-" .. i
		},
	})
end

for i, name in pairs(spitters) do
	data:extend({
		{
			type = "double-setting",
			name = "pitch-" .. name .. "-resistance",
			setting_type = "startup",
			default_value = 10*(i-1)+5,
			minimum_value = 0,
			order = "pitch-a[Evolution]-f[addResistances]-b-" .. i
		},
	})
end

local DayPhase = {'First', 'Second', 'Third', 'Fourth'}

-- Day Phase Length&Brightness
for i, name in pairs(DayPhase) do
	data:extend({
		{
			type = "int-setting",
			name = "pitch-" .. name .. "DayPhaseMaxLength",
			setting_type = "runtime-global",
			default_value = 12-(i*2),
			minimum_value = 1,
			order = "pitch-b[DayLength]-b[MaxDayLength]-" .. i
		},
		{
			type = "int-setting",
			name = "pitch-" .. name .. "DayPhaseMinLength",
			setting_type = "runtime-global",
			default_value = 12-(i*2),
			minimum_value = 1,
			order = "pitch-b[DayLength]-c[MinDayLength]-" .. i
		},
		{
			type = "int-setting",
			name = "pitch-" .. name .. "NightPhaseMaxLength",
			setting_type = "runtime-global",
			default_value = i,
			minimum_value = 1,
			order = "pitch-b[DayLength]-d[MaxDayLength]-" .. i
		},
		{
			type = "int-setting",
			name = "pitch-" .. name .. "NightPhaseMinLength",
			setting_type = "runtime-global",
			default_value = i,
			minimum_value = 1,
			order = "pitch-b[DayLength]-e[MinDayLength]-" .. i
		},
		{
			type = "double-setting",
			name = "pitch-" .. name .. "DayPhaseMaxBrightness",
			setting_type = "runtime-global",
			default_value = 1.0-(i*0.2),
			minimum_value = 0.0,
			maximum_value = 1.0,
			order = "pitch-c[Brightness]-a[MaxDayBrightness]-" .. i
		},
		{
			type = "double-setting",
			name = "pitch-" .. name .. "DayPhaseMinBrightness",
			setting_type = "runtime-global",
			default_value = 1.0-(i*0.2),
			minimum_value = 0.0,
			maximum_value = 1.0,
			order = "pitch-c[Brightness]-b[MinDayBrightness]-" .. i
		},
		{
			type = "double-setting",
			name = "pitch-" .. name .. "NightPhaseMaxBrightness",
			setting_type = "runtime-global",
			default_value = 0.0,
			minimum_value = 0.0,
			maximum_value = 1.0,
			order = "pitch-c[Brightness]-c[MaxDayBrightness]-" .. i
		},
		{
			type = "double-setting",
			name = "pitch-" .. name .. "NightPhaseMinBrightness",
			setting_type = "runtime-global",
			default_value = 0.0,
			minimum_value = 0.0,
			maximum_value = 1.0,
			order = "pitch-c[Brightness]-d[MinDayBrightness]-" .. i
		},
	})
end

data-updates.lua
require 'Libs/Utility/math'

--Increase all pollution
for _, prototype_type in pairs(data.raw) do
    for _, prototype in pairs(prototype_type) do
        if prototype.energy_source then
            if prototype.energy_source.emissions and prototype.energy_source.emissions > 0.001 then
                prototype.energy_source.emissions = prototype.energy_source.emissions * 2
            end
        end

        --Manually buff powerrr as night is more painful
        if prototype.type == "solar-panel" then
            local stringMetric = string.sub(prototype.production, string.len(prototype.production) - 2)
            local stringVal = string.sub(prototype.production, 0, string.len(prototype.production) - 2)
            local numericalVal = tonumber(stringVal)
            numericalVal = (numericalVal * 3) / 10

            prototype.production = numericalVal .. stringMetric
        end
    end
end

data-final-fixes.lua
--local biters = {'small-biter', 'medium-biter', 'big-biter', 'behemoth-biter'}
--local spitters = {'small-spitter', 'medium-spitter', 'big-spitter', 'behemoth-spitter'}
--
--for _, name in pairs(biters) do
--    log(name)
--    log(serpent.block(data.raw["unit"][name].attack_parameters.ammo_type.action.action_delivery.target_effects.damage.amount))
--    log(serpent.block(data.raw["unit"][name].resistances))
--end
--for _, name in pairs(spitters) do
--    log(name)
--    log(serpent.block(data.raw["unit"][name].attack_parameters.ammo_type))
--    log(serpent.block(data.raw["unit"][name].resistances))
--end
--log(serpent.block(data.raw.projectile["acid-projectile-purple"].action.action_delivery.target_effects[3]))
--log(serpent.block(data.raw.wall["stone-wall"].resistances))
local EnemyHealthScale = settings.startup["pitch-EnemyHealthScale"].value
local EnemySwarmScale = settings.startup["pitch-EnemySwarmScale"].value
local BiterDamageScale = settings.startup["pitch-BiterDamageScale"].value
local SpitterDamageScale = settings.startup["pitch-SpitterDamageScale"].value
local EnemyMovementScale = settings.startup["pitch-EnemyMovementScale"].value

--Spawners spawn more and more frequently
for _, prototype in pairs(data.raw["unit-spawner"]) do
    prototype.max_count_of_owned_units = math.ceil(6 * EnemySwarmScale)
    prototype.max_friends_around_to_spawn = math.ceil(8 * EnemySwarmScale)
end

local biters = {'small-biter', 'medium-biter', 'big-biter', 'behemoth-biter'}
local spitters = {'small-spitter', 'medium-spitter', 'big-spitter', 'behemoth-spitter'}

local resistances = {}
for _, name in pairs(biters) do
    resistances[name] = settings.startup["pitch-" .. name .. "-resistance"].value
end
for _, name in pairs(spitters) do
    resistances[name] = settings.startup["pitch-" .. name .. "-resistance"].value
end

--Biters have more HP and join the attack more rapidly if you're polluting too much
data.raw["unit"]["small-biter"].pollution_to_join_attack = math.ceil(133 * (1.0 / EnemySwarmScale))         --Default 200
data.raw["unit"]["small-biter"].max_health = math.ceil(10 * EnemyHealthScale)                               --Default 15
data.raw["unit"]["medium-biter"].pollution_to_join_attack = math.ceil(666 * (1.0 / EnemySwarmScale))            --Default 1000
data.raw["unit"]["medium-biter"].max_health = math.ceil(50 * EnemyHealthScale)                              --Default 75
data.raw["unit"]["big-biter"].pollution_to_join_attack = math.ceil(2666 * (1.0 / EnemySwarmScale))          --Default 4000
data.raw["unit"]["big-biter"].max_health = math.ceil(500 * EnemyHealthScale)                                    --Default 375
data.raw["unit"]["behemoth-biter"].pollution_to_join_attack = math.ceil(6650 * (1.0 / EnemySwarmScale))     --Default 20000
data.raw["unit"]["behemoth-biter"].max_health = math.ceil(2500 * EnemyHealthScale)                          --Default 3000
data.raw["unit"]["small-spitter"].pollution_to_join_attack = math.ceil(133 * (1.0 / EnemySwarmScale))       --Default 200
data.raw["unit"]["small-spitter"].max_health = math.ceil(8 * EnemyHealthScale)                              --Default 10
data.raw["unit"]["medium-spitter"].pollution_to_join_attack = math.ceil(665 * (1.0 / EnemySwarmScale))      --Default 600
data.raw["unit"]["medium-spitter"].max_health = math.ceil(37 * EnemyHealthScale)                                --Default 50
data.raw["unit"]["big-spitter"].pollution_to_join_attack = math.ceil(2666 * (1.0 / EnemySwarmScale))            --Default 1500
data.raw["unit"]["big-spitter"].max_health = math.ceil(375 * EnemyHealthScale)                              --Default 200
data.raw["unit"]["behemoth-spitter"].pollution_to_join_attack = math.ceil(6650 * (1.0 / EnemySwarmScale))   --Default 10000
data.raw["unit"]["behemoth-spitter"].max_health = math.ceil(2000 * EnemyHealthScale)                            --Default 2000

for _, name in pairs(biters) do
    data.raw["unit"][name].attack_parameters.ammo_type.action.action_delivery.target_effects.damage.amount =
        math.ceil(data.raw["unit"][name].attack_parameters.ammo_type.action.action_delivery.target_effects.damage.amount * BiterDamageScale)
    data.raw["unit"][name].movement_speed = data.raw["unit"][name].movement_speed * EnemyMovementScale
    data.raw["unit"][name].distance_per_frame = data.raw["unit"][name].distance_per_frame * EnemyMovementScale
    if settings.startup["pitch-addResistances"].value then
        if not data.raw["unit"][name].resistances then
            data.raw["unit"][name].resistances = {}
        end
        table.insert(data.raw["unit"][name].resistances, {type = "fire", percent = resistances[name]})
    end
end

log("Pitch Black adjusting spitters")
for _, name in pairs(spitters) do
    log(name .. ':')
    log('Original modifier: ' .. serpent.line(data.raw["unit"][name].attack_parameters.damage_modifier))
    data.raw["unit"][name].attack_parameters.damage_modifier = data.raw["unit"][name].attack_parameters.damage_modifier or 1
    data.raw["unit"][name].attack_parameters.damage_modifier =
        math.ceil(data.raw["unit"][name].attack_parameters.damage_modifier * SpitterDamageScale)
    log('New modifier: ' .. serpent.line(data.raw["unit"][name].attack_parameters.damage_modifier))
    data.raw["unit"][name].movement_speed = data.raw["unit"][name].movement_speed * EnemyMovementScale
    data.raw["unit"][name].distance_per_frame = data.raw["unit"][name].distance_per_frame * EnemyMovementScale
    if settings.startup["pitch-addResistances"].value then
        if not data.raw["unit"][name].resistances then
            data.raw["unit"][name].resistances = {}
        end
        table.insert(data.raw["unit"][name].resistances, {type = "fire", percent = resistances[name]})
    end
end

local speed = math.min(SpitterDamageScale, BiterDamageScale)
if speed ~= 1 then
    data.raw["repair-tool"]["repair-pack"].speed = data.raw["repair-tool"]["repair-pack"].speed * speed * 0.9
    --log(string.format("repair speed: %s", data.raw["repair-tool"]["repair-pack"].speed))
end

if settings.startup["pitch-disableSmoke"].value then
    data.raw.stream["flamethrower-fire-stream"].smoke_sources = nil
    data.raw.fire["fire-flame"].on_fuel_added_action = nil
    data.raw.fire["fire-flame"].smoke = nil
    data.raw.fire["fire-flame-on-tree"].smoke = nil
end

data:extend({
    {
        type = "explosion",
        name = "pitch_explosion",
        flags = {"not-on-map"},
        animations = {
            {
                filename = "__Pitch_Black_kor__/trans1.png",
                priority = "extra-high",
                width = 1,
                height = 1,
                frame_count = 1,
                animation_speed = 0.5
            },
            {
                filename = "__Pitch_Black_kor__/trans1.png",
                priority = "extra-high",
                width = 1,
                height = 1,
                frame_count = 1,
                animation_speed = 0.5
            },
            {
                filename = "__Pitch_Black_kor__/trans1.png",
                priority = "extra-high",
                width = 1,
                height = 1,
                frame_count = 1,
                animation_speed = 0.5
            },
            {
                filename = "__Pitch_Black_kor__/trans1.png",
                priority = "extra-high",
                width = 1,
                height = 1,
                frame_count = 1,
                animation_speed = 0.5
            }
        },
        light = {intensity = 0, size = 20, color = {r=1.0, g=1.0, b=1.0}},
        sound =
        {
            aggregation =
            {
                max_count = 1,
                remove = true
            },
            variations =
            {
                {
                    filename = "__Pitch_Black_kor__/scream.ogg",
                    volume = 1
                }
            }
        }
    }
})
